text/0000-raw-references.md

```diff
@@ -0,0 +1,292 @@
+retrieve a pointer to such.
+
+A [similarly motivated RFC](https://github.com/rust-lang/rfcs/pull/2582) exists
+that tries to approach this problem by adding a MIR operation that performs a
+direct pointer to pointer-to-field conversion in a few defined cases. This
+leaves open the question of a formalized approach to the problem.
+
+While exposing such properties through the type system would not be backwards
+compatible, it could still be possible to attach such properties internally.
+Through this system, we can also provide better warnings and give the necessary
+foundation to properly discuss extended guarantees.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Each reference (`&` or `&mut`) will have an internal–that is, not observable by
+rust code–tracking state named `raw`. While a reference is `raw`, it will be
+represented as a pointer in syntax-to-MIR lowering and the value to which
+dereferences need not be valid. The act of unactivating the raw status of a
+reference will be called `settling` for the purpose of this document, how this
+can happen will be discussed below. When a reference with active `raw` status
+is converted to a pointer, this will be called a raw pointer cast and would be
+a no-op in MIR.
+
+```
+// The result of this borrow is a raw reference.
+let x = unsafe { &packed.field as *const _ };
+```
+
+In short, a reference will keep its raw status as long as its pointed-to content
+is not accessed. Since the tracking of references should stay function local,
+passing a reference as an argument to another function or returning it will
+unset the `raw` status. Note that raw references can only originate from unsafe
+code.  Thus, when a raw reference is cast to a pointer in safe code, this is an
+indicator for potentially incorrect usage of unsafe and a warning could be
+emitted.
+
+Since this solution works (invisbly) at type-level, we keep the composability
+properties that code relies on. In particular, we can decompose expressions
+into subexpressions:
+
+```
+let x = unsafe {
+    // Desugared version. `raw` is reference with `raw` status.
+    let raw = &(*packed).field;
+    // Cast with active raw status, raw pointer cast.
+    raw as *const _
+};
+```
+
+This also propagates out of unsafe blocks, even though a raw reference from a
+pointer can only be created within an unsafe block. This proposal explicitely
+does not intends stabilize the guarantee of the following code not invoking
+undefined behaviour but nevertheless proposes to not exploit it in MIR until
+better tools are created in the surface language to differentiate the reference
+requirements.
+
+```
+let x = unsafe { &packed.field };
+let x = x as *const _;
+```
+
+Since this is the only use of `x`, and `x` is never settle but originated in an
+unsafe block, this is a strong code smell. Specifically, it would be strictly
+safer for this reference to never leave the unsafe code block. By tracking the
```

Comment:
This seems to be suggesting that the compiler ascribes significance to the boundaries of an unsafe block, which is untrue.  The unit of unsafety is the entire module that contains an `unsafe` block.

Reply:
because it's at a block boundary, it becomes a true reference value - it's somewhat equivalent to a function return.

Reply:
That would be my interpretation of the other rfc as well. However, in this rfc I intended to keep the `raw` status around for all basic blocks of the current function. Thus, function return makes it a true reference value but the block boundary would not yet invoke undefined behaviour.

Reply:
That's fair, and I don't think that's ureasonable.